### PR TITLE
Write initial thoughts on ADS support in diary

### PR DESCRIPTION
**Pull request by zachmandeville**
This is the start of work to make sure our existing tests work for an ADS configured server, along with writing ADS-only tests.  I wanted to double check my assumptions with team before going too far down this road.

This PR is just an org file documenting my work plan.  Keen to hear what you think!

### CODE REVIEW

**Review by hh — COMMENTED** [review]
I think this is what they have asked for, I'm looking forward to if, when clearly written out this way, is what they wanted.

File: diary/ads-support.org [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+- An xDS server is configured for a single variant on startup
+  + e.g: if it's set to SoTW/ADS, then all CDS,LDS,RDS,EDS features will be handled
+    via ADS
+  + It cannot switch itself from basic to ADS and remain running.
+- The intent of our current tests are the same, no matter if a server is
+  configured as basic or aggregated.
+  + However, there will be additional 'ADS-only' tests
+- A team may want to ensure their server can handle basic /and/ ADS.
+- **The framework should be able to be configured for a server's configuration, a configuration that holds through the entire test suite.**
+- **If a team wants to test their server with multiple configurations, they should run the test suite multiple times, once per configuration.**
```
Comment (hh) [inline diary/ads-support.org:29]: This makes sense, I think this is the first time it's been clear to me that this is really a requirement.

Comment (markdroth) [inline diary/ads-support.org:29]: Instead of requiring users to run the test suite multiple times, I think it would be better for the test suite to be told which variants of the protocol the xDS server supports, and it should automatically run all of the relevant tests against the server for that set of variants.

One possible way to structure this would be to give each test a list of tags to indicate which variant(s) the test is relevant for.  For example, we can have tags for "SotW", "incremental", "aggregated", and "non-aggregated", and each test can have two or more tags (each test will need at least one of "SotW" or "incremental" and at least one of "aggregated" or "non-aggregated").  For each protocol variant that the server supports, the test framework will go through the list of tests and run all tests that are relevant for that variant -- e.g., when testing non-aggregated SotW, it will run all tests that have both "SotW" and "non-aggregated" tags.

File: diary/ads-support.org [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  + It [[https://github.com/ii/xds-test-harness/tree/de750b5ba26ba3f0de5a6ecca1fcac20b787ee30][currently]] only supports SoTW(State of the World)/Basic
+  + SoTW/ADS(Aggregated Discovery) should be the added next, followed by Delta/Basic and Delta/ADS.
+- An xDS server is configured for a single variant on startup
+  + e.g: if it's set to SoTW/ADS, then all CDS,LDS,RDS,EDS features will be handled
+    via ADS
+  + It cannot switch itself from basic to ADS and remain running.
+- The intent of our current tests are the same, no matter if a server is
+  configured as basic or aggregated.
+  + However, there will be additional 'ADS-only' tests
+- A team may want to ensure their server can handle basic /and/ ADS.
```
Comment (markdroth) [inline diary/ads-support.org:27]: A given server may support any arbitrary subset of the protocol variants.  Some possible examples of what a server might support:
- non-aggregated SotW only
- SotW only, both aggregated and non-aggregated (but neither of the incremental variants)
- aggregated only, both SotW and incremental (but neither of the non-aggregated variants)
- all 4 variants
- only aggregated SotW and non-aggregated incremental (but not non-aggregated SotW or aggregated incremental)

That last example is probably fairly unlikely, but there's no reason that the test framework should make any assumptions about it.

File: diary/ads-support.org [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+In addition, the suite would run tests tagged with "@ads". The easiest way to do
+this is through negation. By default the tests run as "run everything BUT ads".
+With the "@ads" flag set it becomes "run everything".
+
+** Process
+*** TODO Double Check assumptions and design with team
+*** Implement hook for switching between ADS and Basic
+*** Refactor functions as necessary
+*** Write ADS only test
+*** Celebrate and Dance
```
Comment (markdroth) [inline diary/ads-support.org:56]: I'm looking forward to this step! :)

File: diary/ads-support.org [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+** Questions
+*** Are my assumptions about ADS and our harness correct?
+*** Should the harness detect how the target is configured?
```
Comment (markdroth) [inline diary/ads-support.org:59]: I don't think it should try to auto-detect, because that could lead to edge cases where the implementor expects that it's supporting a given variant, and the test suite passes, but it did so because it mis-detected and assumed that that variant was not supported, so it didn't actually run the tests for that variant.  That would essentially be a false positive, where the test suite says that the target supports something that it does not actually support.

I think it will be much cleaner to explicitly configure the framework for the set of variants that should be tested for a given target.  Commandline flags are probably fine for now.

**Review by markdroth — COMMENTED** [review]
Thanks for writing this up!  Please let me know if you have any questions about any of this.

**Comment by zachmandeville** [comment]
Thank you for the notes, @markdroth and @hh .  I've incorporated them into a new org file, which I've included in PR #38. I am going to close this one in favour of that one.
